org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/XtextGeneratorResourceSetInitializer.java

```diff
@@ -0,0 +1,152 @@
+import org.eclipse.emf.ecore.resource.ResourceSet;
+import org.eclipse.emf.mwe.utils.GenModelHelper;
+import org.eclipse.emf.mwe.utils.StandaloneSetup;
+import org.eclipse.xtext.resource.IResourceServiceProvider;
+import org.eclipse.xtext.xbase.lib.Procedures.Procedure1;
+
+import com.google.common.collect.Iterables;
+
+/**
+ * Initializes a resource set in order to load the grammar of a language. The
+ * resources to include are specified via
+ * {@link XtextGeneratorLanguage#addReferencedResource(String)}.
+ */
+public class XtextGeneratorResourceSetInitializer {
+	private static final Logger LOG = Logger.getLogger(XtextGeneratorResourceSetInitializer.class);
+
+	public void initialize(ResourceSet resourceSet, List<String> referencedResources) {
+		StandaloneSetup delegate = new StandaloneSetup();
+		delegate.setResourceSet(resourceSet);
+		resourceSet.getPackageRegistry().put(EcorePackage.eNS_URI, EcorePackage.eINSTANCE);
+		referencedResources.forEach((String it) -> {
+			loadResource(it, resourceSet);
+		});
+		registerGenModels(resourceSet);
+		registerEPackages(resourceSet);
+	}
+
+	private void loadResource(String loadedResource, ResourceSet resourceSet) {
+		URI loadedResourceUri = URI.createURI(loadedResource);
+		ensureResourceCanBeLoaded(loadedResourceUri, resourceSet);
+		resourceSet.getResource(loadedResourceUri, true);
+	}
+
+	private void ensureResourceCanBeLoaded(URI loadedResource, ResourceSet resourceSet) {
+		String fileExtension = loadedResource.fileExtension();
+		if (fileExtension != null) {
+			switch (fileExtension) {
+			case "genmodel":
+				GenModelPackage.eINSTANCE.getEFactoryInstance();
+				IResourceServiceProvider resourceServiceProvider = IResourceServiceProvider.Registry.INSTANCE
+						.getResourceServiceProvider(loadedResource);
+				if (resourceServiceProvider == null) {
+					try {
+						Class<?> genModelSupport = Class.forName("org.eclipse.emf.codegen.ecore.xtext.GenModelSupport");
+						Object instance = genModelSupport.getDeclaredConstructor().newInstance();
+						genModelSupport.getDeclaredMethod("createInjectorAndDoEMFRegistration").invoke(instance);
+					} catch (ClassNotFoundException e) {
+						LOG.debug(
+								"org.eclipse.emf.codegen.ecore.xtext.GenModelSupport not found, GenModels will not be indexed");
+					} catch (Exception e) {
+						LOG.error("Couldn't initialize GenModel support.", e);
+					}
+				}
+				return;
+			case "ecore":
+				IResourceServiceProvider resourceServiceProvider2 = IResourceServiceProvider.Registry.INSTANCE
+						.getResourceServiceProvider(loadedResource);
+				if (resourceServiceProvider2 == null) {
+					try {
+						Class<?> ecore = Class.forName("org.eclipse.xtext.ecore.EcoreSupportStandaloneSetup");
+						ecore.getDeclaredMethod("setup", new Class[] {}).invoke(null);
+					} catch (ClassNotFoundException e) {
+						LOG.error("Couldn't initialize Ecore support. Is 'org.eclipse.xtext.ecore' on the classpath?");
```

Comment:
"Couldn't initialize Ecore support. Is 'org.eclipse.xtext.ecore' on the classpath?" -> Couldn't initialize Ecore support. It is possible that the library "org.eclipse.xtext.ecore" is missing from the project classpath.

Reply:
did not want to refactor messages here